core/src/main/scala/cats/syntax/enumerable.scala

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2015 Typelevel
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package cats
+package syntax
+
+import cats.kernel._
+
+trait PartialNextSyntax {
+  implicit final def catsSyntaxPartialNext[A](a: A): PartialNextOps[A] =
+    new PartialNextOps(a)
+}
+
+final class PartialNextOps[A](private val lhs: A) extends AnyVal {
+  def partialNext(implicit A: PartialNext[A]): Option[A] =
+    A.partialNext(lhs)
+}
+
+trait NextSyntax extends PartialNextSyntax {
+  implicit final def catsSyntaxNext[A](a: A): NextOps[A] =
+    new NextOps(a)
+}
+
+final class NextOps[A](private val lhs: A) extends AnyVal {
+  def next(implicit A: Next[A]): A =
+    A.next(lhs)
+}
+
+trait PartialPreviousSyntax {
+  implicit final def catsSyntaxPartialPrevious[A](a: A): PartialPreviousOps[A] =
+    new PartialPreviousOps(a)
+}
+
+final class PartialPreviousOps[A](private val lhs: A) extends AnyVal {
+  def partialPrevious(implicit A: PartialPrevious[A]): Option[A] =
+    A.partialPrevious(lhs)
+}
+
+trait PreviousSyntax extends PartialPreviousSyntax {
+  implicit final def catsSyntaxPrevious[A](a: A): PreviousOps[A] =
+    new PreviousOps(a)
+}
+
+final class PreviousOps[A](private val lhs: A) extends AnyVal {
+  def previous(implicit A: Previous[A]): A =
+    A.previous(lhs)
+}
+
+trait BoundedEnumerableSyntax extends NextSyntax with PreviousSyntax {
+  implicit final def catsSyntaxForBoundedEnumerable[A](a: A): BoundedEnumerableOps[A] =
+    new BoundedEnumerableOps(a)
+}
+
+final class BoundedEnumerableOps[A](private val lhs: A) extends AnyVal {
+  def cycleNext(implicit A0: PartialNext[A], A1: LowerBounded[A]): A =
+    A0.partialNext(lhs).getOrElse(A1.minBound)
+
+  def cyclePrevious(implicit A0: PartialPrevious[A], A1: UpperBounded[A]): A =
+    A0.partialPrevious(lhs).getOrElse(A1.maxBound)
```

Comment:
I am a bit concerned that the syntax class implements its own logic.
There's `BoundedEnumerable` typelclass already that implements `cycleNext` and `cyclePrevious`.
Usually the syntax should only do
```suggestion

  def cycleNext(implicit A: BoundedEnumerable[A]): A =
    A.cycleNext(lhs)

  def cyclePrevious(implicit A: BoundedEnumerable): A =
    A.cyclePrevious(lhs)
```
shouldn't it?

Reply:
...on the other hand, _I think_ I see what you meant: `BoundedEnumerable` is too restrictive for that.
And I think this is because the entire hierarchy is sort of not that well-developed yet. Perhaps, there should be two more typeclasses added:
```scala
// just an arbitrary name, feel free to suggest a better one
trait CycleableNext extends PartialNext with LowerBounded {
  def cycleNext ...
}

Reply:
Actually, there's a couple classes like that: `PartialPreviousUpperBounded` and `PartialNextLowerBounded`, but those do not seem to be well-developed either:
https://github.com/typelevel/cats/blob/bdb1ee029f7967bf2b4532a6597c31667186502a/kernel/src/main/scala-2.13%2B/cats/kernel/EnumerableCompat.scala#L28-L42

I wonder, why does it inherit to `PartialNext` if it does not use anything from the latter?

Reply:
@satorg Yeah, there are some issues here. 

So, first off, RE the syntax classes having logic. I don't want to demand a constraint of `BoundedEnumerable` because someone might define a `PartialNext` and `LowerBounded` as separate definitions and then we won't be able to solve for the instance, e.g. like `Decoder` `Encoder` in circe or `Get` `Put` in Doobie.

Secondly, I think there is a more complete and less restrictive encoding for what `PartialPreviousUpperBounded` and `PartialNextLowerBouneded` are trying to do. I'm prototyping it out here: https://github.com/isomarcte/cats/commit/9debe1c1f439a66f5792d3349471f98fd953af51#diff-805a1fb053663f9d27cfcb8586bc2310cbe9a649366549b8d0106a366774e2e3R45

Keep in mind, that is a rough early draft of what I'm calling `Countable` (should probably just be called `Enumerable`). It can be defined for both finite and infinite types, unlike `PartialPreviousUpperBounded`.

I'm trying to work out how the whole hierarchy should look.

The `cycleNext` and `cyclePrevious` functions are interesting. Technically, we could define these for unbounded (infinite) types as well, they would just be aliases for `next` and `previous`. That is to say, the cycle over an infinite domain, is just a cycle which never wraps. This might be nice if someone wanted to endlessly enumerate over a domain, not caring if they were cycling or just going on forever. :shrug: 

`Countable` (or `Enumerable`) would be really helpful for defining an `Interval` type. The stdlib Range and cats-collection Range is sort of trying to be both a convenient way to enumerate a series of elements and also a true mathematical interval, with the stdlib version leaning a bit more into the former and cats-collection's version the latter. I think that is why we have some ergonomic issues with those types. However if we define `Countable` separately then we can solve the "I want to enumerate over some set of values problem" and having that as a constraint also makes writing `Interval` a lot nicer. Specifically, saying that an `Interval` is something that maps to the natural numbers opens up a whole lot of optimizations and ergonomic benefits.

Sorry, I got a little off topic there. All that `Countable` stuff would be another PR once I've sorted out how I think it might all fit together with the existing classes. For now, I just wanted to get the syntax in place for what is already defined, with minimal constraints.

Reply:
I appreciate you deep involvements into the topic. Actually, I think you already have a suggestion on how the concern can be solved, doesn't you?
https://github.com/typelevel/cats/blob/be74ffc69b91e9e42ee4b144a2bae9f012f352dc/kernel/src/main/scala/cats/kernel/Enumerable.scala#L256-L257